docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,496 @@
+
+After compression occurs, the `Content-Encoding` header will contain the name of the compression algorithm used. It will be appended to the list of content encodings if the header was already present with some value.
+
+No encoding header before compression:
+```http request
+POST /upload-data HTTP/1.1
+Host: example.com
+Content-Type: application/json
+Content-Encoding: gzip
+```
+
+Existing `br` encoding in header before compression:
+```http request
+POST /upload-data HTTP/1.1
+Host: example.com
+Content-Type: application/json
+Content-Encoding: br, gzip 
+```
+
+### Compression Algorithms
+Given that Kotlin is multiplatform there should be multiple implementations of each compression algorithm. Currently gzip is the only supported compression algorithm by the server side and the only one supported for
+the SDK. Only gzip for JVM is supported specifically. This is done by using the following classes from the Java standard library: `GZIPInputStream` and `GZIPOutputStream`. The other KMP targets are not implemented, and are planned to be implemented using the CRT.
+
+New algorithms can be added as they become supported using the compression algorithm interface and by adding them to the list of supported algorithms. The same goes for non JVM KMP targets, these can be added by filling in the `actual` declarations for the target.
+
+#### Interfaces
+This interface is used to represent a compression algorithm.
+
+```kotlin
+public interface CompressionAlgorithm {
+    /**
+     * The ID of the compression algorithm
+     */
+    public val id: String
+
+    /**
+     * The name of the content encoding to be appended to the `Content-Encoding` header.
+     * The [IANA](https://www.iana.org/assignments/http-parameters/http-parameters.xhtml)
+     * has a list of registered encodings for reference.
+     */
+    public val contentEncoding: String
+    
+    /**
+     * Compresses a stream
+     */
+    public fun compress(stream: ByteStream): ByteStream
+}
+```
+
+The `id` is used to match a supported `CompressionAlgorithm` to a requested compression algorithm from the trait. The `contentEncoding` is used to set a value in the `Content-Encoding` header, and the `compress` function is used to compress the request.
+
+#### Compression
+Gzip JVM compression implementation:
+
+ByteStream implementation
+```kotlin
+val byteArrayOutputStream = ByteArrayOutputStream()
+val gzipOutputStream = GZIPOutputStream(byteArrayOutputStream)
+
+gzipOutputStream.write(stream.bytes())
+gzipOutputStream.close()
+
+val compressedBody = byteArrayOutputStream.toByteArray()
+byteArrayOutputStream.close()
+
+return compressedBody
+```
+
+The SourceStream and ChannelStream implementation wrap the source so that it compresses into gzip format with each read.
+
+SourceStream implementation
```

Comment:
nit: I don't think the design doc should have a copy of the entire implementation, we can just read the code for that